client/session.go

```diff
@@ -153,18 +159,23 @@ func (s *session) Open() error {
 	s.seriesIteratorsPool.Init()
 	s.Unlock()
 
-	currTopologyMap := s.topo.GetAndSubscribe(s.topoMapCh)
 	if err := s.setTopologyMap(currTopologyMap); err != nil {
 		s.Lock()
 		s.state = stateNotOpen
 		s.Unlock()
 		return err
 	}
+	s.topoWatch = topologyWatch
```

Comment:
The `s.topoWatch = topologyWatch` assignment should be in a lock, also if we're moving the `GetAndSubscribe()` into the lock above all the state setting, I would move the the `if err := setTopologyMap(currTopologyMap); err != nil { return err }` directly below it.